### PR TITLE
fix(dag-node): ensure links are always DAGLinks

### PR DESCRIPTION
**Pull request by dignifiedquire**

### CODE REVIEW

**Comment by daviddias** [comment]
If I'm seeing correctly, this PR has two changes:
- adds a reconstructions of the DAGLinks - This means that if a DAGLink was passed, now the reference that the rest of the program has, is different from the internal to the DAGNode. Can we check to see if it is a DAGLink and not reconstructed? 
- moves the code from function to classes - I believe this is more of a style preference, if we move to classes, it would be better to have the whole module using classes, so that is consistent and more enjoyable to read.

**Comment by dignifiedquire** [comment]
How do suggest checking for the passed values being `DAGLinks`? We can not use `instanceof` and we need to ensure a valid `toJSON` method plus all props are there, but one of them might be null so a simple check for all of them being defined does not work either.

---

The change to using `class` was purely motivated by this issue

``` js
function DAGNode (args) {

  this.someMethod(args[0])
  // throws this.someMethod is not defined

  this.someMethod = () => {}
}
```

I'm okay with moving everything in this module to use `class` just didn't want to do that as that would be a purely personal style oriented change :)

**Comment by daviddias** [comment]
> How do suggest checking for the passed values being DAGLinks?

We can have a method `isDAGLink` on DAGLink for the check. We can also check that the hash that DAGLink receives is a valid multihash, that the name is a string and the size a number.

**Comment by dignifiedquire** [comment]
> We can also check that the hash that DAGLink receives is a valid multihash, that the name is a string and the size a number.

We could, but I have seen a couple of instances in js-ipfs where a `DAGLink` is constructed with just a `name` property, e.g. when calling `object.patch.rmLink` as that method only needs the `name` to remove a link. So adding validation for that would stop us from doing that.

**Comment by daviddias** [comment]
Let's check if not undefined, if it is the right type. That gives a lil more guards (as someone could put a String hash and miss why it was failing) but not make it too restrictive.

**Comment by dignifiedquire** [comment]
So you mean something like 

``` js
if (mh != null && isValidMultihash(mh)) {
 // I am happy and I know it
} else {
  // I am not so happy
}
```

**Comment by dignifiedquire** [comment]
@diasdavid what shall we do with the `class` problem?

**Comment by daviddias** [comment]
👍 to https://github.com/vijayee/js-ipfs-merkle-dag/pull/35#issuecomment-219400649 (with `!==`) and on the `class` problem: The whole module should be consistent, if there is a technical limitation and a good reason to use classes in one type, let's use them in the other ones too

File: src/dag-node.js [hunk 1]
```diff
@@ -1,77 +1,116 @@
...
 const stable = require('stable')
-const bs58 = require('bs58')
+const fs = require('fs')
+const path = require('path')
+const mh = require('multihashes')
+
+const util = require('./util')
+const DAGLink = require('./dag-link')
+
+const proto = protobuf(fs.readFileSync(path.join(__dirname, 'merkledag.proto')))
```
Comment (daviddias) [inline src/dag-node.js:12]: 👍

**Comment by daviddias** [comment]
LGTM. Waiting for CI, then merging and releasing :)

**Comment by daviddias** [comment]
released :)

**Pull request merged**